accepted/2023/feature-switch-attributes.md

```diff
@@ -0,0 +1,718 @@
+
+  ```csharp
+  class RequiresWithVersionAttribute : Attribute {
+      public RequiresWithVersionAttribute(string version) {}
+  }
+
+  class RequiresFooVersionAttribute : RequiresWithVersionAttribute {
+      public RequiresFooVersionAttribute(string version) : base(version) {}
+  }
+
+  class Foo {
+      [FeatureSwitch(typeof(RequiresFooVersion))]
+      public static bool IsSupportedWithVersionAtLeast(string version) => return VersionIsLessThanOrEquals(version, "2.0");
+
+      [RequiresFooVersion("2.0")]
+      public static void Impl_2_0() {
+          // Do some work
+      }
+
+      [RequiresFooVersion("1.0")]
+      public static void Impl_1_0() {
+        // Breaking change was made in version 2.0, where this API is no longer supported.
+          throw new NotSupportedException();
+      }
+  }
+  ```
+
+  Code that was originally built against the 1.0 version, and broken on the upgrade to the 2.0 version, could then be updated with a feature check like this:
+  ```csharp
+  if (Foo.IsSupportedWithVersionAtLeast("2.0")) {
+      Foo.Impl_2_0();
+  } else {
+      Foo.Impl_1_0();
+  }
+  ```
+
+  Although it's not clear in practice where this would be useful. This is not meant as a realistic example.
+
+  The platform compatibility analyzer represents version ranges via a combination of attributes as described in [advanced scenarios for attribute combinations](https://learn.microsoft.com/dotnet/standard/analyzers/platform-compat-analyzer#advanced-scenarios-for-attribute-combinations) (in addition to representing combinations of support or lack of support for various platforms). This can encode a supported or unsupported version range for a given platform, which might alternately be encoded by a single attribute that takes starting and ending versions for support. In any case, the model seems neatly extensible to version numbers should we need them.
+
+- Feature attribute schemas
+
+  We could consider unifying this model with the platform compatibility analyzer. One difference is that the `SupportedOSPlatformAttribute` takes a string indicating the platform name. We would likely need to extend the understanding of feature attributes to support treating "features" differently based on this string, effectively supporting feature attributes which define not a single feature, but a schema that allows representing a class of features. For example:
+
+  ```csharp
+  class OSPlatformAttribute : FeatureAttribute {
+      private protected OSPlatformAttribute([FeatureName] string platformName)
+      {
+          PlatformName = platformName;
+      }
+
+      [FeatureName]
+      public string PlatformName { get; }
+  }
+  ```
+
+  In this example, `FeatureName` indicates that the instances of `SupportedOSPlatformAttribute` should be differentiated based on the value of this parameter.
+
+## Alternate API shapes
+
+### Feature switches without feature attributes
```

Comment:
Personally I like another design: Have a separate type which represents the feature alone - and acts just as metadata.

```csharp
[FeatureName("System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported")]
[FeatureRequirement(typeof(RequiresDynamicCode))]
static class DynamicCodeSupportedFeature {
}

class RequiresDynamicCode : Attribute { }

class RuntimeFeature {
    [FeatureSwitch(typeof(DynamicCodeSupportedFeature))]
    public static bool IsDynamicCodeSupported => // ...

    [FeatureGuard(typeof(DynamicCodeSupportedFeture))]
    public static bool IsDynamicCodeCompiled => // ..
}
```

The current proposal basically collapses the `RequiresDynamicCodeAttribute` with the `DynamicCodeSupportedFeature`. The disadvantage of that is:
* Features which don't have a requirement attribute would still need to define one - which is confusing

Another consideration is visibility. Whichever type we use to "define the feature" (either its own type, or the requires attribute) - if that type is not public, it's not possible to define custom feature guard properties for the feature. This might be useful, but could also become a problem.